cmd/agent/common/autodiscovery.go

```diff
@@ -0,0 +1,116 @@
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+func setupAutoDiscovery(confSearchPaths []string, metaScheduler *scheduler.MetaScheduler) *autodiscovery.AutoConfig {
+	ad := autodiscovery.NewAutoConfig(metaScheduler)
+	ad.AddConfigProvider(providers.NewFileConfigProvider(confSearchPaths), false, 0)
+
+	// Autodiscovery cannot easily use config.RegisterOverrideFunc() due to Unmarshalling
+	var discoveredProviders []config.ConfigurationProviders
+	var discoveredListeners []config.Listeners
+	if config.Datadog.GetBool("autoconf_from_environment") {
+		discoveredProviders, discoveredListeners = discoverAutodiscoveryComponents()
+	}
+
+	// Register additional configuration providers
+	var CP []config.ConfigurationProviders
+	err := config.Datadog.UnmarshalKey("config_providers", &CP)
+	if err == nil {
+		// Add extra config providers
+		for _, name := range config.Datadog.GetStringSlice("extra_config_providers") {
+			CP = append(CP, config.ConfigurationProviders{Name: name, Polling: true})
+		}
+
+		for _, provider := range discoveredProviders {
+			CP = append(CP, provider)
+		}
+	} else {
+		log.Errorf("Error while reading 'config_providers' settings: %v", err)
+	}
+
+	// Adding all found providers
+	for _, cp := range CP {
+		factory, found := providers.ProviderCatalog[cp.Name]
+		if found {
+			configProvider, err := factory(cp)
+			if err == nil {
+				pollInterval := providers.GetPollInterval(cp)
+				if cp.Polling {
+					log.Infof("Registering %s config provider polled every %s", cp.Name, pollInterval.String())
+				} else {
+					log.Infof("Registering %s config provider", cp.Name)
+				}
+				ad.AddConfigProvider(configProvider, cp.Polling, pollInterval)
+			} else {
+				log.Errorf("Error while adding config provider %v: %v", cp.Name, err)
+			}
+		} else {
+			log.Errorf("Unable to find this provider in the catalog: %v", cp.Name)
+		}
+	}
+
+	var listeners []config.Listeners
+	err = config.Datadog.UnmarshalKey("listeners", &listeners)
+	if err == nil {
+		// Add extra listeners
+		for _, name := range config.Datadog.GetStringSlice("extra_listeners") {
+			listeners = append(listeners, config.Listeners{Name: name})
+		}
+
+		for _, listener := range discoveredListeners {
+			listeners = append(listeners, listener)
+		}
+
+		ad.AddListeners(listeners)
+	} else {
+		log.Errorf("Error while reading 'listeners' settings: %v", err)
+	}
+
+	return ad
+}
+
+func discoverAutodiscoveryComponents() ([]config.ConfigurationProviders, []config.Listeners) {
+	detectedProviders := []config.ConfigurationProviders{}
+	detectedListeners := []config.Listeners{}
+
+	// When using automatic discovery of providers/listeners
+	// We automatically activate the environment listener
+	detectedListeners = append(detectedListeners, config.Listeners{Name: "environment"})
+
+	if config.IsFeaturePresent(config.Docker) {
+		detectedProviders = append(detectedProviders, config.ConfigurationProviders{Name: "docker", Polling: true, PollInterval: "1s"})
```

Comment:
we might want to keep the 1s polling interval configurable

Reply:
Then you need to use a custom configuration. That's also why `detectedProviders` are added last. AD will pick up the first occurence if provider/listener registered multiple times (so the custom one).